---
title: "Transcriptomics and the analysis of RNA-Seq data"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Transcriptomics and the analysis of RNA-Seq data

Set up Bioconductor DESeq2
```{r eval=FALSE}
install.packages("BiocManager")
BiocManager::install()
BiocManager::install("DESeq2")
```



Obtain data for today's class
```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
```


Q1. How many genes are in this dataset? 38694
```{r}
head(counts)
nrow(counts)
```

How many experiments?
```{r}
ncol(counts) - 1
```



Q2. How many 'control' cell lines do we have? 
```{r}
head(metadata)
nrow(metadata) /2
```


Lets make sure the metaaata id col matcches rhe colnames of counts
```{r}
colnames(counts)[-1]
```
```{r}
metadata$id
```
```{r}
all(colnames(counts)[-1] == metadata$id )
```


## Analysis : compare the control to drug treated

First we need to access the columns of our countData that are control and then find their mean.
```{r}
metadata$dex=="control"
```

```{r}
control.id <- metadata [metadata$dex=="control", ]$id
control.id
```

Focus on the counts for control experiments
```{r}
head( counts[,control.id] )
```


Mean counts for control experiments 

```{r}
control.mean <- rowSums(counts[,control.id])/length(control.id)
names(control.mean) <- counts$ensgene
```






Now for the treated columns

```{r}
treated.id <- metadata [metadata$dex=="treated", ]$id
treated.id
```

Focus on the counts for the treated experiments 

```{r}
head( counts[,treated.id])
```


Mean counts for treated experiments 

```{r}
treated.mean <- rowSums(counts[,treated.id])/length(treated.id)
names(treated.mean) <- counts$ensgene
```


Let's store these together 

```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

Plot control vs treated!
```{r}
plot(meancounts$control.mean, meancounts$treated.mean)
```


Good case for a log scaled plot

```{r}
plot(meancounts$control.mean, meancounts$treated.mean, log="xy")
```



Fold change

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

There are a couple of "weird" results. Namely, the NaN ("not a number") and -Inf (negative infinity) results


The NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. It turns out that there are a lot of genes with zero expression. Let's filter out 


```{r}
# Test for finding zero entries

x <- c(5,8,0,5)
x==0
which(x==0)
x
```


```{r}
x <- data.frame(happy= c(5,6,0,0), sad=c(0,5,5,0))
x==0
```

```{r}
which(x==0)
```

Arr.ind function tells you the coordinate row,col position of each value that matches your which function
```{r}
which(x==0, arr.ind=TRUE)
```

Focus on the rows only that have zero entries
Unique function tell you which rows have a zero entry 
```{r}
unique( which(x==0, arr.ind=TRUE)[,1])
```


```{r}
inds <- unique( which(x==0, arr.ind=TRUE)[,1])
x[-inds,]
```


Now to do this to our data - exclude all values that have a zero for either control or treated column....
```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)
to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
nrow(mycounts)
```

A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let's filter the dataset both ways to see how many genes are up or down-regulated


```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
sum(up.ind) + sum(down.ind)
```



## Annotation 

We will use Bioconductor's annotation packages to help with ampping various ID schemes to each other. Here we load the AnnotationDbi packages and the annotation package org.Hs.es.db.

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

See what is available in org.Hs

```{r}
columns(org.Hs.eg.db)
```



We can use the mapIds() function to add individual columns to our results table. We provide the row names of our results table as a key, and specify that keytype=ENSEMBL. The column argument tells the mapIds() function which information we want, and the multiVals argument tells the function what to do if there are multiple possible values for a single input value. Here we ask to just give us back the first one that occurs in the database.

```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
```

```{r}
head(mycounts)
```






## DESeq2 Analysis

```{r}
library(DESeq2)
```

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```

```{r}
dds <- DESeq(dds)
```


```{r}
res <- results(dds)
res
```



## Volcano Plot

```{r}
plot(res$log2FoldChange, res$padj)
```


```{r}
plot( res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
```


Make this plot less ull

```{r}
mycols <- rep("gray", length(res$padj))
mycols[ abs(res$log2FoldChange) > 2 ] = "blue"
mycols[ res$padj < 0.05 ] = "red"
plot( res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)",
      col = mycols)
```































